src/chipper/keyboard2.cljs

```diff
@@ -0,0 +1,233 @@
+;; as a name in the absence of a distinct insert mode)
+;; :special ; which is then its own thing, like {:action :octave-up :parameters 10}
+;; :custom  ; or maybe `:macro`? something to facilitate more complex things
+;; then a big handler decides which little handler actually alters state
+;; big handler would basically just be a another map like
+;; {:position  set-position!
+;;  :attribute set-attribute!  ; this needs a better name than attribute, maybe
+;;  etc}
+
+;; Available properties are
+;; :position :frame :mode :attr :macro :special
+;; Available macro types are
+;; :simple   (a sequence of position/frame/mode/attr)
+;; :advanced (a callable?)
+;; and macros mappings are defined as
+;; {:Trigger [:macro {:type type :value value}]}
+;; See edit mode :Backspace for an example
+
+(def measure-size 4)
+
+;; XXX These should maybe be somwhere else, like core, or... something
+;; also note that if you use these, the max index of the cursor is 1 less
+;; because position is 0-indexed
+(def frame-count 32)
+(def frame-length 32)
+(def chan-count 4)
+(def attr-count 3)
+;; a hack, because of course
+(def total-attr-positions (* chan-count attr-count))
+
+(def -motions
+  {:down  [ 1  0  0]
+   :up    [-1  0  0]
+   :left  [ 0  0 -1]
+   :right [ 0  0  1]})
+
+(def normal-keymap
+  {;; relative movement
+   ;; keeping :position for clarity. The representation for cursor position may
+   ;; change in the future, thus -motions
+   :KeyJ        [:relative-position (:down  -motions)]
+   :KeyK        [:relative-position (:up    -motions)]
+   :KeyH        [:relative-position (:left  -motions)]
+   :KeyL        [:relative-position (:right -motions)]
+
+   :KeyEnter    [:relative-position (:down  -motions)]
+   :ArrowDown   [:relative-position (:down  -motions)]
+   :ArrowUp     [:relative-position (:up    -motions)]
+   :ArrowLeft   [:relative-position (:left  -motions)]
+   :ArrowRight  [:relative-position (:right -motions)]
+
+   :KeyW        [:relative-position [measure-size     0 0]]
+   :KeyB        [:relative-position [(- measure-size) 0 0]]
+
+   :ShiftBracketRight [:frame  1]
+   :ShiftBracketLeft  [:frame -1]
+
+   ;; absolute movement
+   :KeyG        [:absolute-position [:first :first :first]]
+   :ShiftKeyG   [:absolute-position [:last  :last  :last]]
+   :Digit0      [:absolute-position [:first :first :first]]
+   :ShiftDigit4 [:absolute-position [:last  :last  :last]]
+
+   ;; mode change
+   :KeyI [:mode :edit]
+   :KeyV [:mode :visual]
+
+   ;; other
+   :KeyX        [:macro {:type :simple :value [[:attr nil] (:down-line -motions)]}]
+   ;; so these could be special or just, like... :tempo, and :octave
+   ;; I'm leaving them out for now
+   ;; :Period [:special [:tempo 1]]
+   ;; :ShiftPeriod [:special [:tempo -1]]
+   :Space [:special :play-pause]})
+
+(def edit-keymap
+  ;; hm, this isn't going to work quite like you want it to.
+  ;; this is fine for inputting notes...
+  {:KeyA [:attr :C]    :KeyW [:attr :C#]
```

Comment:
Representing note names as keywords makes sense, but it also makes stuff like `C-a` and any interesting chord operations harder to implement. It'll be better to just use ints...

Reply:
t̨h̢e͝ pact̨ i̶s se͏a̕le҉d͞

Reply:
probably should be one of the last changes to make, since it changes the representation for state used elsewhere in the app. gonna be a fun one

oof that annoys me. ~you wanna take it and learn many things you didn't want to know about `ui`?~

Reply:
yeah I can take it

Reply:
#31

Reply:
[commitmaxxing]